File: src_old/AMBPatcher/ShaChecker.cs
using System;
using System.IO;
using System.Collections.Generic;
using System.Security.Cryptography;

namespace AMBPatcher
{
    public static class ShaChecker
    {
        public static HashAlgorithm SHAcsp = SHA1.Create();
        public static string Sha(byte[] file)
        {
            var hash = SHAcsp.ComputeHash(file);
            return BitConverter.ToString(hash).Replace("-", "");
        }

        public static void ShaRemove(string file_name)
        {
            string orig_file_sha_root = Path.Combine("mods_sha", file_name);

            if (Directory.Exists(orig_file_sha_root))
            {
                var sha_files = Directory.GetFiles(orig_file_sha_root, "*", SearchOption.AllDirectories);

                foreach (string file in sha_files)
                    File.Delete(file);
            }
        }

        public static bool ShaChanged(string file_name, List<string> mod_files, List<string> mod_paths)
        {
            if (!Settings.SHACheck) return true;

            bool files_changed = false;

            List<string> sha_list = new() { };

            string orig_file_sha_root = Path.Combine("mods_sha", file_name);

            if (Directory.Exists(orig_file_sha_root))
                sha_list = new List<string>(Directory.GetFiles(orig_file_sha_root, "*.txt", SearchOption.AllDirectories));

            //Checking SHA1s
            for (int i = 0; i < mod_files.Count; i++)
            {
                if (files_changed) { break; }

                string mod_file_full = Path.Combine("mods", mod_paths[i], mod_files[i]);
                string mod_file_sha = Path.Combine("mods_sha", mod_files[i] + ".txt");

                if (sha_list.Contains(mod_file_sha))
                    sha_list.Remove(mod_file_sha);
                else
                {
                    files_changed = true;
                    break;
                }

                if (File.Exists(mod_file_sha))
                {
                    string sha_tmp = Sha(File.ReadAllBytes(mod_file_full));
                    if (sha_tmp != File.ReadAllText(mod_file_sha)) { files_changed = true; }
                }
                else files_changed = true;
            }

            //Checking if there're removed files
            //And removing those SHAs
            if (sha_list.Count > 0)
            {
                files_changed = true;

                foreach (string file in sha_list)
                    File.Delete(file);
            }

            return files_changed;
        }

        public static void ShaWrite(string relative_mod_file_path, string full_mod_file_path)
        {
            string sha_file = Path.Combine("mods_sha", relative_mod_file_path + ".txt");
            string sha_dir = Path.GetDirectoryName(sha_file);

            Directory.CreateDirectory(sha_dir);
            File.WriteAllText(sha_file, Sha(File.ReadAllBytes(full_mod_file_path)));
        }

    }
}

File: src/amb-rs/src/sha_checker.rs
use std::{fs, path::{Path, PathBuf}};
use glob::glob;
use sha1::{Digest, Sha1};

pub fn get(data: impl AsRef<[u8]>) -> String {
    Sha1::digest(data).iter().map(|x| format!("{:02x}", x)).collect()
}

pub fn remove(file_name: String)
{
    let orig_file_sha_root = Path::new("mods_sha").join(file_name);

    if orig_file_sha_root.is_dir() {
        fs::remove_dir_all(orig_file_sha_root).expect("Bad thing happened #123");
    }
}

pub fn is_changed(do_sha_check: bool, file_name: &String, mod_files: &Vec<String>, mod_paths: &Vec<String>) -> bool {
    if !do_sha_check {
        return true;
    }

    let mut files_changed = false;

    let mut sha_list = Vec::<String>::new();

    let orig_file_sha_root = Path::new("mods_sha").join(file_name);

    if orig_file_sha_root.is_dir() {
        let glob_result = glob(&(orig_file_sha_root.display().to_string() + "/**/*.txt"));
        match glob_result {
            Ok(glob_result) => {
                for entry in glob_result {
                    match entry {
                        Ok(path) => sha_list.push(path.display().to_string()),
                        Err(e) => println!("Glob error: {e}"),
                    }
                }
            },
            Err(e) => println!("Error: {e}"),
        }
    }

    //Checking SHA1s
    let mut i = 0;
    while i < mod_files.len() {
        if files_changed { 
            break;
        }

        let mod_file_full = Path::new("mods").join(mod_paths[i].clone()).join(mod_files[i].clone());
        let mod_file_sha = Path::new("mods_sha").join(mod_files[i].clone() + ".txt");

        match sha_list.iter().position(|x| x.to_string() == mod_file_sha.display().to_string()) {
            Some(index) => sha_list.remove(index),
            None => {
                files_changed = true;
                break;
            }
        };

        if mod_file_sha.is_file() {
            let sha_tmp = get(fs::read(mod_file_full).expect("Bad, bad thing happened"));
            if sha_tmp != fs::read_to_string(mod_file_sha).expect("Badder thing happened") {
                files_changed = true;
            }
        } else {
            files_changed = true;
        }
        i += 1;
    }

    //Checking if there're removed files
    //And removing those SHAs
    if sha_list.len() > 0
    {
        files_changed = true;

        for file in sha_list {
            fs::remove_file(file).expect("Worse thing happened");
        }
    }

    return files_changed;
}

pub fn write(relative_mod_file_path: String, full_mod_file_path: PathBuf) {
    let sha_file = Path::new("mods_sha").join(relative_mod_file_path + ".txt");
    let sha_dir = sha_file.parent().unwrap();

    fs::create_dir_all(sha_dir.parent().unwrap()).expect("Failed to create directory for SHA-1 file");
    fs::write(sha_file, get(fs::read(full_mod_file_path).expect("Failed to read file for SHA-1 file "))).expect("Couldn't write SHA-1 of file");
}